rust/common/error.rs

```diff
@@ -140,54 +100,151 @@ macro_rules! max {
-   VARIABLE_OUT_OF_SCOPE_MATCH = 16: "The variable '{}' is out of scope of the match query.",
-   VARIABLE_OUT_OF_SCOPE_DELETE = 17: "The deleted variable '{}' is out of scope of the match query.",
-   NO_VARIABLE_IN_SCOPE_INSERT = 18: "None of the variables in 'insert' ('{}') is within scope of 'match' ('{}')",
-   VARIABLE_NOT_NAMED = 19: "Anonymous variable encountered in a match query filter.",
-   INVALID_VARIABLE_NAME = 20: "The variable name '{}' is invalid; variables must match the following regular expression: '^[a-zA-Z0-9][a-zA-Z0-9_-]+$'.",
-   MISSING_CONSTRAINT_RELATION_PLAYER = 22: "A relation variable has not been provided with role players.",
-   INVALID_CONSTRAINT_PREDICATE = 25: "The '{}' constraint may only accept a string value as its operand, got '{}' instead.",
-   INVALID_CONSTRAINT_DATETIME_PRECISION = 26: "Attempted to assign DateTime value of '{}' which is more precise than 1 millisecond.",
-   INVALID_DEFINE_QUERY_VARIABLE = 27: "Invalid define/undefine query. User defined variables are not accepted in define/undefine query.",
-   INVALID_UNDEFINE_QUERY_RULE = 28: "Invalid undefine query: the rule body of '{}' ('when' or 'then') cannot be undefined. The rule must be undefined entirely by referring to its label.",
-   INVALID_RULE_WHEN_MISSING_PATTERNS = 29: "Rule '{}' 'when' has not been provided with any patterns.",
-   INVALID_RULE_WHEN_NESTED_NEGATION = 30: "Rule '{}' 'when' contains a nested negation.",
-   INVALID_RULE_THEN = 31: "Rule '{}' 'then' '{}': must be exactly one attribute ownership, or exactly one relation.",
-   INVALID_RULE_THEN_HAS = 32: "Rule '{}' 'then' '{}' tries to assign type '{}' to variable '{}', but this variable already had a type assigned by the rule 'when'. Try omitting this type assignment.",
-   INVALID_RULE_THEN_VARIABLES = 33: "Rule '{}' 'then' variables must be present in the 'when', outside of nested patterns.",
-   INVALID_RULE_THEN_ROLES = 34: "Rule '{}' 'then' '{}' must specify all role types explicitly or by using a variable.",
-   REDUNDANT_NESTED_NEGATION = 35: "Invalid query containing redundant nested negations.",
-   VARIABLE_NOT_SORTED = 36: "Variable '{}' does not exist in the sorting clause.",
-   INVALID_COUNT_VARIABLE_ARGUMENT = 38: "Aggregate COUNT does not accept a Variable.",
-   ILLEGAL_GRAMMAR = 39: "Illegal grammar: '{}'",
-   ILLEGAL_CHAR_IN_LABEL = 40: "'{}' is not a valid Type label. Type labels must start with a letter, and may contain only letters, numbers, '-' and '_'.",
+error_messages! { TypeQLError
+    code: "TQL", type: "TypeQL Error",
+    SyntaxErrorDetailed(usize, String, String, String) =
+        3: "There is a syntax error at line {}:\n{}\n{}\n{}",
+    InvalidCasting(&'static str, &'static str, &'static str, &'static str) =
+        4: "Enum '{}::{}' does not match '{}', and cannot be unwrapped into '{}'.",
+    MissingPatterns() =
+        5: "The query has not been provided with any patterns.",
```

Comment:
We now return an enum variant wrapping format arguments. This cleans up the point of use: `SYNTAX_ERROR_DETAILED.format(&[&line.to_string(), ...])` becomes `TypeQLError::SyntaxErrorDetailed(line, ...)`, and also allows the compiler to verify the types of the arguments and the number of the arguments to the format string. This approach also unlocks full capabilities of built-in Rust formatting, though we aren't using it anywhere yet.